Support using netrc for non-proxy HTTP credentials

yuvipanda opened the pull request:
<!-- Thank you for your contribution! -->

## What do these changes do?

Currently, setting `trust_env=True` in `ClientSession` will only use the `.netrc`
file if present for authentication to proxies. With this change, the credentials will
also be used for regular HTTP hosts, and Basic Auth will be sent if an appropriate
entry is present in the user's `.netrc` file.

## Are there changes in behavior for the user?

YES! Currently, their .netrc file will be ignored when not making proxy requests. This
will start sending any credentials in netrc files to hosts.

The behavior of this PR seems to have possibly been the suggested behavior when
netrc support was initially added, based on this comment: https://github.com/aio-libs/aiohttp/pull/2584#issuecomment-348935249.

The behavior with this PR applied also matches requests and other libraries. In fact, this
was a deep dive output after trying to figure out why accessing NASA open data with requests
works but aiohttp does not :D  (https://github.com/nsidc/earthaccess/issues/188 has the trail crumbs).

If this is considered a breaking change, I'm happy to instead introduce a *new* param that
toggles this behavior instead. However, `trust_env` is what requests and other libraries use,
so may be ok.

## Related issue number

https://github.com/aio-libs/aiohttp/issues/2581
https://github.com/aio-libs/aiohttp/pull/2584#issuecomment-348935249
(outside aiohttp) https://github.com/nsidc/earthaccess/issues/188

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

yuvipanda commented:
I will add tests and documentation changes soon! Wanted to open this early to get @asvetlov's opinion (and others' too maybe?) on wether this is a breaking change or not.

THANK YOU SO MUCH for all the work on aiohttplib! It is heavily used in the open science / climate science ecosystem, and we are grateful for it.

Dreamsorcerer commented on aiohttp/client_reqrep.py:
```diff
@@ -428,11 +430,23 @@ def update_transfer_encoding(self) -> None:
+                    if auth_from_netrc:
+                        *logins, password = auth_from_netrc
+                        login = logins[0] if logins[0] else logins[-1]
+                        auth = BasicAuth(cast(str, login), cast(str, password))
```
Why the cast()? It seems like both of these can be `None`, if `BasicAuth` doesn't allow `None`, then we shouldn't try to pass them here.

yuvipanda replied:
@Dreamsorcerer ah, I was just copying code from how this is handled in the proxies (in helpers.py), I will take a slightly deeper look onto why the casts are there in the code.

Dreamsorcerer replied:
Ah, I see, looks like asvetlov was trying to get typing added across the entire codebase, so probably didn't look into it in detail. Probably worth fixing it as it suggests a bug...

If this code comes from there, I also wonder if it's worth deduplicating it.

yuvipanda replied:
@Dreamsorcerer I've deduplicated it now! I found the type signature for netrc authenticators at https://github.com/python/typeshed/blob/04d4cbd92f2f6b2f845e1c3e0b257a96c3e170a5/stdlib/netrc.pyi#L13, will try to figure out how to type it correctly here shortly

yuvipanda replied:
@Dreamsorcerer so I fell into a rabbithole, and it actually turns out this behavior changed between 3.10 and 3.11! https://github.com/python/cpython/pull/26330 changed when something returns None vs empty string too.

yuvipanda replied:
I'll dig in some more and figure out the appropriate thing to do, and leave a lot of comments as well :)

yuvipanda replied:
Here is some sample code with different netrc files:

```python
import netrc
import tempfile

cases = [
    # Host, with login and password
    "machine testhost login login1 password pass1",
    # Host, with login, account and password
    "machine testhost login login1 account account1 password pass1",
    # Host, with account and password
    "machine testhost account login2 password pass2",
    # Host, with just password, no account nor login
    "machine testhost password pass2",
    # Default with just login, no password
    "default account login3",
]

for line in cases:
    with tempfile.NamedTemporaryFile() as f:
        f.write(line.encode())
        f.flush()
        n = netrc.netrc(f.name)
        print(n.authenticators("testhost"))
```

Here is the output for python3.10:

```
('login1', None, 'pass1')
('login1', 'account1', 'pass1')
('', 'login2', 'pass2')
('', None, 'pass2')
Traceback (most recent call last):
  File "/Users/yuvipanda/code/aiohttp/test_netrc.py", line 21, in <module>
    n = netrc.netrc(f.name)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/netrc.py", line 31, in __init__
    self._parse(file, fp, default_netrc)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/netrc.py", line 82, in _parse
    raise NetrcParseError(
netrc.NetrcParseError: malformed default entry default terminated by '' (/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/tmp8a66t8qr, line 1)
```

Here is the output for 3.11:

```
('login1', '', 'pass1')
('login1', 'account1', 'pass1')
('', 'login2', 'pass2')
('', '', 'pass2')
('', 'login3', '')
```

ignoring the exception, the basic difference seems to be that until 3.10, the type of any of those fields is really `Optional[str]`, while with 3.11 the type is just `str`. Given we want to be compatible, I'm going to treat everything as if it is `Optional[str]`, but set None values to '' to match 3.11 behavior.

yuvipanda replied:
@Dreamsorcerer ok I've updated the code to be clearer now! Let me know what you think :)

I'll also add test cases based on the helper script I pasted above soon.

Dreamsorcerer replied:
Typeshed fix: https://github.com/python/typeshed/pull/9376

Dreamsorcerer commented on aiohttp/client_reqrep.py:
```diff
@@ -428,11 +431,20 @@ def update_transfer_encoding(self) -> None:
+                netrc_obj = netrc_from_env()
+                if netrc_obj:
+                    auth = basicauth_from_netrc(netrc_obj, self.url.host)
+                    if auth:
+                        self.headers[hdrs.AUTHORIZATION] = auth.encode()
```
I'd be inclined not to repeat this code here. Instead doing the above code under `if auth is None and trust_env:` and having another `if auth is None: return` block afterwards...

Dreamsorcerer commented on aiohttp/client_reqrep.py:
```diff
@@ -428,11 +431,20 @@ def update_transfer_encoding(self) -> None:
         """Set basic auth."""
         if auth is None:
             auth = self.auth
         if auth is None:
+            if trust_env:
+                # If no auth is specified, we read from netrc.
+
+                netrc_obj = netrc_from_env()
+                if netrc_obj:
+                    auth = basicauth_from_netrc(netrc_obj, self.url.host)
+                    if auth:
+                        self.headers[hdrs.AUTHORIZATION] = auth.encode()
+
             return
```
```suggestion
        if auth is None and trust_env:
            netrc_obj = netrc_from_env()
            if netrc_obj:
                auth = basicauth_from_netrc(netrc_obj, self.url.host)
        if auth is None:
            return
```

Maybe this?

Dreamsorcerer commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,31 @@ class ProxyInfo:
+
+        auth = BasicAuth(username, password)
+        return auth
+
```
```suggestion
    return None

```
If a function should return a value, then it should be explicit.

Dreamsorcerer replied:
In fact, would be even nicer to read if we just put:
```
if auth_from_netrc is None:
    return None
```
And then unindented the rest of the code.

Dreamsorcerer commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,31 @@ class ProxyInfo:
+    auth_from_netrc = netrc_obj.authenticators(host)
+    if auth_from_netrc is not None:
+        login, account, password = auth_from_netrc
+        if not login:
+            # login is provided, we just use it as username.
```
```suggestion
        if login:
            # login is provided, we just use it as username.
```

Think that's reversed?

Dreamsorcerer commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,31 @@ class ProxyInfo:
+        else:
+            # neither login nore account are provided, we use blank username
+            # this matches python 3.11 behavior
+            username = ""
```
login is always `str`, right?
So, we could just simplify this whole block to `username = login if login else account`

Dreamsorcerer replied:
Or
```
if not login and account:
    login = account
```
And just use `login` below.

Dreamsorcerer replied:
OK, thinking about the typeshed change, I'd like to go with:
```
# TODO(PY311): username = login or account
username = login if login or account is None else account
```

The `None` check will get highlighted as a mypy error once we drop Python 3.10 support.

Dreamsorcerer commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,31 @@ class ProxyInfo:
+            # this matches python 3.11 behavior
+            username = ""
+
+        if password is None:
```
This will also nicely get caught by mypy once 3.10 is dropped.

Dreamsorcerer commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,31 @@ class ProxyInfo:
+            # this matches python 3.11 behavior
+            password = ""
+
+        auth = BasicAuth(username, password)
+        return auth
```
```suggestion
        return BasicAuth(username, password)
```

yuvipanda commented:
@Dreamsorcerer I really appreciate all the code clarity suggestions, thank you. I've implemented all of them, and tried to keep clarifying comments as well.

I've also added three unit tests. I'm not sure if `test_netrc_from_env` is redundant, happy to remove it if you think so.

yuvipanda commented:
I also appreciate the checklist in the PR body template, and believe I have ticked all the boxes.

yuvipanda commented:
I do not know what the correct version this is to be added in. I'd love to see this released sooner than later (:D) - would that be 3.8.4 or 3.9 or 4.x? It probably is either 3.9 or 4.x, but I am not sure what the next release is to be...

webknjaz commented:
3.9 if bacoported.

yuvipanda commented:
@webknjaz ah perfect! i've modified the PR to say 3.9, and would love for it to be backported :)

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1232,47 @@ def test_loose_cookies_types(loop: Any) -> None:
+        f.write(netrc_contents.encode())
+        f.flush()
+        try:
+            # save and restore NETRC env variable if already set
```
Instead of this, use the built-in `monkeypatch` fixture that will unpatch the env automatically after the test function finishes.

yuvipanda replied:
TIL about https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.MonkeyPatch.setenv, which is super useful! Thanks for pointing it out :)

webknjaz commented on tests/test_helpers.py:
```diff
@@ -975,3 +977,51 @@ def test_populate_with_cookies():
     assert parse_http_date(value) == expected
+
+
+def test_netrc_from_env():
```
Same here — use built-in fixtures plz.

yuvipanda replied:
Yep, done. I didn't know pytest shipped with somehing for tmp_path!

I'll do the same for the mock.

yuvipanda commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1232,47 @@ def test_loose_cookies_types(loop: Any) -> None:
+    trust_env: bool,
+    expected_auth: Optional[helpers.BasicAuth],
+):
+    with tempfile.NamedTemporaryFile() as f:
```
ooooh, TIL about https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html! never knew it before, thanks for pointing it out :)

yuvipanda commented:
Thank you, @webknjaz! I think I've addressed the issues you pointed out, and learnt a couple new things about pytest along the way!

Dreamsorcerer commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,46 @@ def test_loose_cookies_types(loop: Any) -> None:
+):
+
+    netrc_file_path = tmp_path / ".netrc"
+    with open(netrc_file_path, "w") as f:
```
```suggestion
    with netrc_file_path.open("w") as f:
```

Dreamsorcerer commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,46 @@ def test_loose_cookies_types(loop: Any) -> None:
+):
+
+    netrc_file_path = tmp_path / ".netrc"
+    with open(netrc_file_path, "w") as f:
+        f.write(netrc_contents)
+        f.flush()
```
```suggestion
    netrc_file_path.write_text(netrc_contents)
```

Dreamsorcerer replied:
Also, the flush appears to be redundant in the original code, as you are closing the file when you exit the `with` block.

Dreamsorcerer commented on tests/test_helpers.py:
```diff
@@ -975,3 +976,55 @@ def test_populate_with_cookies():
+
+def test_netrc_from_env(tmp_path: pathlib.Path, monkeypatch: pytest.MonkeyPatch):
+    netrc_file_path = tmp_path / ".netrc"
+    with open(netrc_file_path, "w") as f:
+        f.write("machine example.com login username password pass\n")
+        f.flush()
```
```suggestion
    netrc_file_path.write_text("machine example.com login username password pass\n")
```

Dreamsorcerer commented on tests/test_helpers.py:
```diff
@@ -975,3 +976,55 @@ def test_populate_with_cookies():
+    expected_auth: helpers.BasicAuth,
+):
+    netrc_file_path = tmp_path / ".netrc"
+    with open(netrc_file_path, "w") as f:
+        f.write(netrc_contents)
+        f.flush()
```
```suggestion
    netrc_file_path.write_text(netrc_contents)
```

yuvipanda commented:
TIL about `write_text` too! Thank you, @Dreamsorcerer :)

webknjaz reviewed (commented):
I revised this again, dedicating more attention to interfaces and abstraction layers. I feel like it could be refactored to make the expectations more straightforward. Take a look at the comments.

yuvipanda commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,28 @@ class ProxyInfo:
     proxy_auth: Optional[BasicAuth]
 
 
+def basicauth_from_netrc(netrc_obj: netrc.netrc, host: str) -> Optional[BasicAuth]:
+    auth_from_netrc = netrc_obj.authenticators(host)
+    if auth_from_netrc is None:
+        return None
```
My heuristic for exceptions vs returning none is to get a sense for which is the 'exceptional' circumstance. I would say that *not* finding either a netrc file or an entry for a host is the most common circumstance, even when trust_env is true, while finding an entry is the exceptional circumstance. This is why I prefer it returning None vs exception. There is also a performance penalty here, that is paid for every request when `trust_env` is true. https://switowski.com/blog/ask-for-permission-or-look-before-you-leap/ quantifies this a little, and also suggests the same heuristic I am suggesting - both conceptually and from a performance perspective.

Python 3.11 makes this a bit better with zero cost exceptions (https://bugs.python.org/issue40222) but that only applies if an exception is not actually raised. In this case, as an exception will be raised most of the time so we do not benefit from that optimization.

So my suggestion is to leave this as is. However, given y'all are the maintainers, I will happily fold and implement the suggested changes if you think that is the better way to go :)

yuvipanda replied:
In the interest of making this easier for you to review, I've implemented your suggestions as a PR to this PR: https://github.com/yuvipanda/aiohttp/pull/2. Merging that will update this PR to use exceptions for control flow. I think it is a better fit, conceptually and from a perf perspective (almost half an ms per request!) to not use exceptions here and leave it as is. But I'm totally happy to be told otherwise, and will merge that PR into this PR :)

yuvipanda replied:
@Dreamsorcerer i would prefer to keep the `None` if it doesn't bother you each way, as even if they have `trust_env` on, the `netrc` file is unlikely to have an entry for *most* of the hosts requests are being sent to - just some along the way, as it is just Basic Auth. However, I'll wait for @webknjaz's opinion too!

webknjaz replied:
I don't think that microoptimizing this is valuable in this case. In Python, EAFP is preferred because it values human readability. This is one of the main motivations of using it in the first place — if you do want to write something high-performance, you'd write it in something like Rust and expose it as a Python C-extension. So for the maintainers, readability/maintainability is often more important than performance, too. The optimization shouldn't be premature and should usually target known/measured bottlenecks. Without any proof that this specific bit is comparable in performance with other activities the framework does, there's no saying that it'll noticably change anything globally. aiohttp has optimizations in places that matter — parsing HTTP traffic, for example.
How often does this happen, anyway? Once per `TCPConnector` creation == once per host? Only when the app sets `trust_env=True` explicitly? Does this mean that if it's set, it's pretty much expected to exist more often than not?
I'd argue that it'd be better to look into caching than microoptimizations. But then, again, it'd be up to the caller/end-user to implement such caching as well as wrap the client object initialization, maybe?

yuvipanda replied:
@webknjaz I've merged https://github.com/yuvipanda/aiohttp/pull/2 into this PR now, so it uses exceptions for control flow as you had suggested :)

yuvipanda replied:
I can provide a longer description of the performance impact, but honestly I don't think it particularly matters - the 'is this the exceptional circumstance or expected?' is my usual criteria and performance is just a side benefit. I hope the current changes meet with your approval :)

yuvipanda commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,28 @@ class ProxyInfo:
+    # Upto python 3.10, account could be None if not specified,
+    # and login will be empty string if not specified. From 3.11,
+    # login and account will be empty string if not specified.
+    # If login is present, we prefer that over account as username
+    username = login if (login or account is None) else account
```
Something about this rubbed me the wrong way, so I dug in some more. The initial code, which used `account` as *username* if `login` is empty, came from requests (https://github.com/psf/requests/blob/ac3be98b19f4d09c6a970b271a3ae30f3d0858f7/requests/utils.py#L238).

However, I actually believe requests is *wrong* here. If you look at the [inetutils netrc docs](https://www.gnu.org/software/inetutils/manual/html_node/The-_002enetrc-file.html), it says:

>  Supply an additional account password. If this token is present, the auto-login process will supply the specified string if the remote server requires an additional account password, or the auto-login process will initiate an ACCT command if it does not.

So it seems to be used as an additional **password**, not as username. I looked up what `ACCT` is, and it seems to be the FTP ACCT command (https://cr.yp.to/ftp/user.html), allowing users to have multiple 'subaccounts' of sorts inside a single login. However, note that even with FTP, you can not use 'ACCT' independent of 'LOGIN'.

I looked at some other clients - curl doesn't even recognize `account` (https://everything.curl.dev/usingcurl/netrc for docs, https://github.com/curl/curl/blob/07dfbc08bfb4322b257d6fae775451ec58e63c80/lib/netrc.c#L65 for code). Neither does httpx (https://github.com/encode/httpx/blob/e5bc1ea533aa5cfcc4f7b179e9faa27f689ed91f/httpx/_utils.py#L152). 

So my suggestion here is to actually *not* follow requests, but instead to follow curl and just ignore the `account` field entirely, as sort of being irrelevant to the HTTP protocol. This would also affect existing users (although I doubt it) though, so if you consider it a breaking change, am happy to submit it as a separate PR that can go into 4.x. I am also happy to keep the status quo - although the digging here was mostly because I realized I could ask myself 'wait, *why* are we prefering login?'.

yuvipanda replied:
I have created a PR to my PR here that implements this: https://github.com/yuvipanda/aiohttp/pull/1 :) I would prefer for it to go in as a *separate* PR once this gets merged, and not get backported to 4.0. However, I'm also happy to include that her as is, or just let status quo be (and implement the suggested change here). Your call :)

Dreamsorcerer replied:
I'm not against dropping it, if it looks like it's never going to be used. It doesn't look like it's useful to use account as a fallback to user...

webknjaz replied:
> and not get backported to 4.0.

`master` is `4.0` FWIW. Did you mean not backporting it into `3.9`? If yes, maybe a deprecation warning or sorts would still be useful there.

yuvipanda replied:
@webknjaz yep my bad I meant not backporting into 3.9.

I've accepted your suggestion here. What I'd love is for this PR to get merged and get backported to 3.9, and then I'll make a separate PR that removes the 'account' and applies a `DeprecationWarning`. We can decide in that PR what versions it should be in. Hope that is ok?

yuvipanda commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,44 @@ def test_loose_cookies_types(loop: Any) -> None:
+    monkeypatch.setenv("NETRC", str(netrc_file_path))
+
+    req = make_request("get", f"http://{hostname}", trust_env=trust_env)
+    if expected_auth:
```
This makes sense, I'll implement this.

yuvipanda replied:
Done!

yuvipanda commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,44 @@ def test_loose_cookies_types(loop: Any) -> None:
+    expected_auth: Optional[helpers.BasicAuth],
+):
+
+    netrc_file_path = tmp_path / ".netrc"
+    netrc_file_path.write_text(netrc_contents)
+
+    monkeypatch.setenv("NETRC", str(netrc_file_path))
```
TIL about indirect and usefixtures! I'll figure out how they work and split these tests to use a fixture!

yuvipanda replied:
Done!

yuvipanda commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,44 @@ def test_loose_cookies_types(loop: Any) -> None:
+    trust_env: bool,
+    expected_auth: Optional[helpers.BasicAuth],
+):
+
```
Done

yuvipanda commented on tests/test_helpers.py:
```diff
@@ -975,3 +976,51 @@ def test_populate_with_cookies():
+
+
+def test_netrc_from_env(tmp_path: pathlib.Path, monkeypatch: pytest.MonkeyPatch):
+    netrc_file_path = tmp_path / ".netrc"
+    netrc_file_path.write_text("machine example.com login username password pass\n")
+
+    monkeypatch.setenv("NETRC", str(netrc_file_path))
```
Done, and parametrized.

yuvipanda commented on aiohttp/client_reqrep.py:
```diff
@@ -428,10 +431,14 @@ def update_transfer_encoding(self) -> None:
             auth = self.auth
+        if auth is None and trust_env:
+            netrc_obj = netrc_from_env()
+            if netrc_obj:
+                auth = basicauth_from_netrc(netrc_obj, self.url.host)
```
See https://github.com/aio-libs/aiohttp/pull/7131/files#r1054794474

yuvipanda replied:
But TIL about `contextlib.suppress` which is pretty cool!

yuvipanda commented on aiohttp/helpers.py:
```diff
@@ -261,16 +282,8 @@ def proxies_from_env() -> Dict[str, ProxyInfo]:
-                *logins, password = auth_from_netrc
-                login = logins[0] if logins[0] else logins[-1]
-                auth = BasicAuth(cast(str, login), cast(str, password))
+                auth = basicauth_from_netrc(netrc_obj, proxy.host)
```
See https://github.com/aio-libs/aiohttp/pull/7131/files#r1054794474

yuvipanda commented:
Thank you so much for your thorough review, @webknjaz!

yuvipanda commented:
From https://github.com/aio-libs/aiohttp/actions/runs/3729041540/jobs/6325371951, there is also a mypy warning that seems relevant-ish.

```
aiohttp/client_reqrep.py:441: error: Argument 2 to "basicauth_from_netrc" has incompatible type "Optional[str]"; expected "str"  [arg-type]
```

I investigated this in a little more detail, and it is because `self.url.host` is `Optional[str]` - `yarl` sets it to `None` if the URL is relative (https://yarl.aio-libs.org/en/latest/api.html#yarl.URL). However, `aiohttp` already handles this by throwing an exception if we try to make a request to a relative URL (as that does not make sense in this context):

```
           aiohttp.client_exceptions.InvalidURL: /relative-url
```

Perhaps a `typing.cast` is appropriate here, to tell mypy that we *know* `self.url.host` will not be None? I personally never like `typing.cast` but not sure what exactly to do here.

Dreamsorcerer commented:
> Perhaps a `typing.cast` is appropriate here, to tell mypy that we _know_ `self.url.host` will not be None? I personally never like `typing.cast` but not sure what exactly to do here.

What if you subclass `ClientRequest` and use relative urls? Not sure if that makes sense or not, just a thought worth investigating...

The actual implementation would work with `None` anyway, the netrc code just does `host in dict`. Otherwise, I'd just put an `assert self.url.host is not None`.

yuvipanda commented:
@webknjaz idk if having a relative URL in a subclass is supported or even makes sense, as am not sure how a client will figure out *where* to connect if there is no hostname / ip. But I agree this 'just works' because of the None. I added an explicit `is not None` check, because regardless of relative URL support, it does not make sense to consult netrc if hostname is not set

yuvipanda commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,28 @@ class ProxyInfo:
 
 
+def basicauth_from_netrc(netrc_obj: netrc.netrc, host: str) -> Optional[BasicAuth]:
+    auth_from_netrc = netrc_obj.authenticators(host)
```
Yep, I don't want to touch too much in this one PR!

webknjaz replied:
Agreed.

webknjaz commented on docs/client_reference.rst:
```diff
@@ -179,6 +179,11 @@ The client session supports the context manager protocol for self closing.
 
       Get proxy credentials from ``~/.netrc`` file if present.
 
+      Get HTTP Basic Auth credentials from ``~/.netrc`` file if present.
+
+      If ``NETRC`` environment variable is set, read from file specified
+      there rather than from ``~/.netrc``
```
```suggestion
      Get HTTP Basic Auth credentials from :file:`~/.netrc` file if present.

      If :envvar:`NETRC` environment variable is set, read from file specified
      there rather than from :file:`~/.netrc`.
```

(proper markup + trailing period)

webknjaz replied:
The `:envvar:` role may generate a warning in Sphinx, though. In this case, you may need to either ignore it in `conf.py`, or declare it.

webknjaz commented on docs/client_reference.rst:
```diff
@@ -189,6 +194,10 @@ The client session supports the context manager protocol for self closing.
 
+      .. versionchanged:: 3.9
+
+         Added support for reading HTTP Basic Auth credentials from ``~/.netrc`` file.
```
Looks like you marked this as resolved by mistake.

webknjaz replied:
This is still relevant.

webknjaz commented on tests/conftest.py:
```diff
@@ -1,6 +1,7 @@
 # type: ignore
 import asyncio
 import os
+import pathlib
```
This seems legit.

webknjaz commented on tests/conftest.py:
```diff
@@ -197,3 +198,23 @@ def selector_loop() -> None:
+
+@pytest.fixture
+def netrc_contents(
+    tmp_path: pathlib.Path,
```
```suggestion
    tmp_path: Path,
```

webknjaz commented on tests/conftest.py:
```diff
@@ -197,3 +198,23 @@ def selector_loop() -> None:
+    netrc_file_path = tmp_path / ".netrc"
+    netrc_file_path.write_text(netrc_contents)
+
+    monkeypatch.setenv("NETRC", str(netrc_file_path))
```
It could be useful to return the path too, so it could be used for populating the contents:
```suggestion
    monkeypatch.setenv("NETRC", str(netrc_file_path))

    return netrc_file_path
```

webknjaz commented on tests/test_helpers.py:
```diff
@@ -975,3 +975,59 @@ def test_populate_with_cookies():
+)
+@pytest.mark.usefixtures("netrc_contents")
+def test_basicauth_from_netrc(
+    netrc_contents: str,
```
This is incorrect: `netrc_contents` currently returns `NoneType` and not `str`. But with my suggestion above, it'll be `pathlib.Path`.

OTOH, since you're not using it in the test function, you don't need to list it as an arg at all — `@pytest.mark.usefixtures("netrc_contents")` takes care of the injection. But if you end up using it, you won't need `@pytest.mark.usefixtures("netrc_contents")`.

webknjaz commented on tests/test_helpers.py:
```diff
@@ -975,3 +975,59 @@ def test_populate_with_cookies():
+    """Test that netrc file contents are properly parsed into BasicAuth tuples"""
+    netrc_obj = helpers.netrc_from_env()
+
+    if expected_auth:
+        assert expected_auth == helpers.basicauth_from_netrc(netrc_obj, hostname)
```
The test functions are supposed to be simple. Having them branchy is a clear signal that you're trying to squash two distinct tests into one. Currently, the pytest test report will show `test_basicauth_from_netrc` but what you're actually testing for is (1) `test_basicauth_present_in_netrc` and (2) `test_read_basicauth_from_missing_netrc`.

yuvipanda replied:
GitHub had automatically marked this as resolved before, I am not sure why. I've split this up into two tests named after your suggestions

webknjaz commented on tests/test_helpers.py:
```diff
@@ -975,3 +975,59 @@ def test_populate_with_cookies():
+    expected_auth: helpers.BasicAuth,
+):
+    """Test that netrc file contents are properly parsed into BasicAuth tuples"""
+    netrc_obj = helpers.netrc_from_env()
```
Looks like this could be a module-local fixture.

yuvipanda replied:
Given that it is only used twice, I'd *love* to keep it as is. My brain already hurts from the amount of indirection going on here with the parametrization, I'd love to let this one be.

webknjaz commented on tests/test_helpers.py:
```diff
@@ -975,3 +975,59 @@ def test_populate_with_cookies():
+    if expected_auth:
+        assert expected_auth == helpers.basicauth_from_netrc(netrc_obj, hostname)
+    else:
+        with pytest.raises(LookupError):
```
`pytest.raises()` should always use the `match=` arg. Otherwise, the test is too narrow and there could be uncaught false-positives appearing over time.

yuvipanda replied:
Done

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
+@pytest.mark.usefixtures("netrc_contents")
+def test_basicauth_from_netrc_absent(
+    make_request: Any,
+    netrc_contents: str,
```
(same here)

yuvipanda replied:
Removed

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
+    netrc_contents: str,
+    hostname: str,
+    trust_env: bool,
+    expected_auth: Optional[helpers.BasicAuth],
```
This seems unused.

yuvipanda replied:
Removed

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
+    netrc is not empty.
+    """
+    req = make_request("get", f"http://{hostname}", trust_env=trust_env)
+    if trust_env:
```
Does this mean that the function does not test anything when `trust_env == False`?

webknjaz replied:
This is still relevant.

yuvipanda replied:
It now tests that no Authorization header is sent

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
+def test_basicauth_from_netrc_present(
+    make_request: Any,
+    netrc_contents: str,
+    hostname: str,
```
This doesn't need to be a param because it's always the same.

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
         assert _gen_default_accept_encoding() == expected
+
+
+@pytest.mark.parametrize(
+    ["netrc_contents", "hostname", "trust_env", "expected_auth"],
+    [
+        (
+            "machine example.com login username password pass\n",
+            "example.com",
+            True,
+            helpers.BasicAuth("username", "pass"),
+        ),
+        (
+            "machine example.com login username password pass\n",
+            "example.com",
+            False,
+            None,
+        ),
+    ],
+    indirect=("netrc_contents",),
+)
```
This can be simplified a bit
```suggestion
@pytest.mark.parametrize(
    ("trust_env", "expected_auth"),
    (
        pytest.param(
            True,
            helpers.BasicAuth("username", "pass"),
            id='reading netrc requested',
        ),
        pytest.param(
            False,
            None,
            id='reading netrc disabled',
        ),
    ),
)
@pytest.mark.parametrize(
    "netrc_contents",
    (
        "machine example.com login username password pass\n",
    ),
    indirect=("netrc_contents",),
)
```

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
+        assert req.headers[hdrs.AUTHORIZATION] == expected_auth.encode()
+
+
+@pytest.mark.parametrize(
+    ["netrc_contents", "hostname", "trust_env", "expected_auth"],
+    [
+        ("", "example.com", True, None),
+        ("", "example.com", False, None),
+    ],
+    indirect=("netrc_contents",),
+)
```
```suggestion
@pytest.mark.parametrize("trust_env", (True, False))
@pytest.mark.parametrize(
    "netrc_contents",
    ("", ),
    indirect=("netrc_contents",),
)
```

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
+    indirect=("netrc_contents",),
+)
+@pytest.mark.usefixtures("netrc_contents")
+def test_basicauth_from_netrc_absent(
```
So the file isn't missing, but empty? Why not just say that?

yuvipanda replied:
Renamed test case

webknjaz commented on tests/conftest.py:
```diff
@@ -197,3 +198,23 @@ def selector_loop() -> None:
+    netrc_contents = request.param
+
+    netrc_file_path = tmp_path / ".netrc"
+    netrc_file_path.write_text(netrc_contents)
```
It may be useful to skip file creation when parametrize is not used or is set to `None`:
```suggestion
    if netrc_contents is not None:
        netrc_file_path.write_text(netrc_contents)
```

yuvipanda commented:
Thanks again for the detailed review, @webknjaz! I will try to address them all tonight.

webknjaz commented on aiohttp/client_reqrep.py:
```diff
@@ -428,10 +432,14 @@ def update_transfer_encoding(self) -> None:
         """Set basic auth."""
         if auth is None:
             auth = self.auth
+        if auth is None and trust_env and self.url.host is not None:
```
let's label the condition:
```suggestion
        URL_HAS_A_HOST = self.url.host is not None
        if auth is None and trust_env and URL_HAS_A_HOST:
```

webknjaz replied:
I don't remember if there are any tests for sockets, but it'd be interesting to check if we support HTTP queries to UNIX file sockets as well as abstract sockets. This is the only case I can think of not having a hostname. OTOH, theoretically `file://` URLs could fit this too.

yuvipanda replied:
@webknjaz `aiohttp` client doesn't seem to support `file` URLs anyway. Regardless, I hope it's ok to leave this as is for now and not dig too much deeper.

webknjaz replied:
I realized that maybe URL for unix and abstract sockets would be another use-case. Although, we probably don't test if that works...
Anyway, let's keep it as is for now.

Dreamsorcerer commented:
> @webknjaz idk if having a relative URL in a subclass is supported or even makes sense, as am not sure how a client will figure out _where_ to connect if there is no hostname / ip. But I agree this 'just works' because of the None. I added an explicit `is not None` check, because regardless of relative URL support, it does not make sense to consult netrc if hostname is not set

Seems like some people are finding a use for it. #7142

webknjaz commented:
> Seems like some people are finding a use for it. #7142

That issue does not seem legit, though.

yuvipanda commented:
Just wanted to pop back and say I'll come back to this later this week / early next week!

webknjaz commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,35 @@ class ProxyInfo:
 
+def basicauth_from_netrc(netrc_obj: netrc.netrc, host: str) -> BasicAuth:
+    """
+    Return BasicAuth tuple with credentials for given host from netrc_obj.
```
```suggestion
    Return :py:class:`~aiohttp.BasicAuth` tuple with credentials for given ``host`` from ``netrc_obj``.
```

webknjaz commented on aiohttp/helpers.py:
```diff
@@ -244,6 +243,35 @@ class ProxyInfo:
+    """
+    Return BasicAuth tuple with credentials for given host from netrc_obj.
+
+    Raises LookupError if netrc_obj is None or if no entry is found for the host.
```
```suggestion
    :raises LookupError: if ``netrc_obj`` is :py:data:`None` or if no entry is found for the ``host``.
```

webknjaz commented on docs/client_reference.rst:
```diff
@@ -179,6 +179,11 @@ The client session supports the context manager protocol for self closing.
 
+      Get HTTP Basic Auth credentials from :file:`~/.netrc` file if present.
+
+      If :envvar:`NETRC` environment variable is set, read from file specified
```
It looks like https://aiohttp--7131.org.readthedocs.build/en/7131/glossary.html needs an entry for the env var to be linkable. It's currently unlinked: https://aiohttp--7131.org.readthedocs.build/en/7131/client_reference.html#aiohttp.ClientSession.

yuvipanda replied:
I've added an `..envvar` directive there, but it is now at the bottom of the page rather than in line. I can instead switch the `:envvar:` here to `:term:` and just put `NETRC` into the glossary if you would prefer

webknjaz commented on tests/conftest.py:
```diff
@@ -197,3 +198,26 @@ def selector_loop() -> None:
+    request: pytest.FixtureRequest,
+):
+    """
+    Prepare a ``.netrc`` file with given contents.
```
```suggestion
    Prepare :file:`.netrc` with given contents.
```

webknjaz commented on tests/conftest.py:
```diff
@@ -197,3 +198,26 @@ def selector_loop() -> None:
+    """
+    Prepare a ``.netrc`` file with given contents.
+
+    monkeypatches the NETRC environment variable to point to created
+    file.
```
```suggestion
    Monkey-patches :envvar:`NETRC` to point to created file.
```

webknjaz commented on tests/conftest.py:
```diff
@@ -197,3 +198,26 @@ def selector_loop() -> None:
+    monkeypatches the NETRC environment variable to point to created
+    file.
+    """
+    netrc_contents = request.param
```
```suggestion
    netrc_contents = getattr(request, 'param', None)
```

webknjaz replied:
(apparently, `.param` isn't there if it's not set)

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
+@pytest.mark.usefixtures("netrc_contents")
+def test_basicauth_from_netrc_present(
+    make_request: Any,
+    netrc_contents: str,
```
This is unused.
```suggestion
```

webknjaz commented on tests/test_client_request.py:
```diff
@@ -1230,3 +1230,61 @@ def test_loose_cookies_types(loop: Any) -> None:
+            id="reading netrc requested",
+        ),
+        pytest.param(
+            False,
```
It still looks like this is an entirely different test case.

yuvipanda replied:
I'll split it up. I think this was one of your suggested edits initially though :)

yuvipanda replied:
Split

webknjaz commented on tests/test_helpers.py:
```diff
@@ -975,3 +975,59 @@ def test_populate_with_cookies():
+    indirect=("netrc_contents",),
+)
+@pytest.mark.usefixtures("netrc_contents")
+def test_netrc_from_env(netrc_contents: str, hostname: str, expected_username: str):
```
It's unused
```suggestion
def test_netrc_from_env(hostname: str, expected_username: str):
```

webknjaz commented:
@yuvipanda looks like there's a number of unsolved comments that need to be processed before we proceed.

yuvipanda commented:
@webknjaz yep, I managed to only get to *some* of these. I have some more time scheduled for this PR this week.

yuvipanda commented:
I believe I've addressed everything. You might still have opinions on https://github.com/aio-libs/aiohttp/pull/7131#discussion_r1071693726 and https://github.com/aio-libs/aiohttp/pull/7131/files/4eadd65dcf102cd523aa40812fb1fafe61e4e4a4#r1071666671 @webknjaz

yuvipanda commented:
I see two relevant errors perhaps in `make mypy`:

```
aiohttp/helpers.py:273: error: Argument 1 to "BasicAuth" has incompatible type "Optional[str]"; expected "str"  [arg-type]
aiohttp/client_reqrep.py:443: error: Argument 2 to "basicauth_from_netrc" has incompatible type "Optional[str]"; expected "str"  [arg-type]
```

Both of these are things we guard against with an `if` and an *intermediate variable*, but mypy doesn't seem to detect that. I've this PR to my PR open that fixes these https://github.com/yuvipanda/aiohttp/pull/3.

However, these intermediate bools were added at your explicit request, @webknjaz (see https://github.com/aio-libs/aiohttp/pull/7131#discussion_r1053328611). I personally find what mypy want, without intermediate boolean variables, a lot more readable as well. However, at this juncture I'll basically do whatever it is that you want to get this PR merged - so if you can point me to a way to make mypy happy while also keeping your intermediate variables I'll do so :)

Dreamsorcerer commented:
> However, these intermediate bools were added at your explicit request, @webknjaz (see [#7131 (comment)](https://github.com/aio-libs/aiohttp/pull/7131#discussion_r1053328611)). I personally find what mypy want, without intermediate boolean variables, a lot more readable as well.

Yeah, I don't think the second one improves readability at all. The first one is also scheduled for removal in a few years anyway, so I personally am not bothered about keeping them.

yuvipanda commented:
@Dreamsorcerer i've hence merged in https://github.com/yuvipanda/aiohttp/pull/3 so mypy should pass now!

yuvipanda commented:
yay, all tests seem to pass now!

yuvipanda commented:
Just wanted to check-in on this again, @Dreamsorcerer and @webknjaz :)

Dreamsorcerer commented on docs/glossary.rst:
```diff
@@ -153,3 +155,16 @@
       https://pypi.python.org/pypi/yarl
+
+
+Enviornment Variables
```
```suggestion
Environment Variables
```

yuvipanda replied:
Whoops, thanks for catching this, @Dreamsorcerer

yuvipanda commented on tests/test_helpers.py:
```diff
@@ -975,3 +975,66 @@ def test_populate_with_cookies():
+    indirect=("netrc_contents",),
+)
+@pytest.mark.usefixtures("netrc_contents")
+def est_basicauth_present_in_netrc(
```
```suggestion
def test_basicauth_present_in_netrc(
```

yuvipanda commented:
The windows test failures seem unrelated? And the missing code coverage result is due to differences in python versions, so I hope that's ok.

yuvipanda commented:
This time there's a macos test failure :) So i'd imagine those are flakes.

the codecov warning is one missed line, and that's due to differing behavior in versions of python. I hope that's ok to ignore.

Dreamsorcerer commented:
> This time there's a macos test failure :) So i'd imagine those are flakes.

Not flaky, they just don't work yet. If check has passed, then that means the tests are passing correctly.

> the codecov warning is one missed line, and that's due to differing behavior in versions of python. I hope that's ok to ignore.

Hmm, that's not right. We run on Python 3.7-3.11, so if there was test coverage, it would have run. However, that code doesn't seem important and will be removed in a future version, so I'm not going to bother investigating.

Dreamsorcerer commented:
If @webknjaz is happy to resolve his comments, then I'll merge it in.

yuvipanda commented:
> We run on Python 3.7-3.11, so if there was test coverage, it would have run.

Ah, hmm. I'm not sure why that's happening then :|



> However, that code doesn't seem important and will be removed in a future version, so I'm not going to bother investigating.

I agree wholeheartedly!

yuvipanda commented:
I'm going on vacation from tuesday for 15 days, and wanted to check-in and see if there is anything I can do before I go :)

webknjaz commented:
> However, at this juncture I'll basically do whatever it is that you want to get this PR merged - so if you can point me to a way to make mypy happy while also keeping your intermediate variables I'll do so :)

I suppose you're talking about this [`524c04d` (#7131)](https://github.com/aio-libs/aiohttp/pull/7131/commits/524c04d61a3a4384f314df23bbd5445604678fc7). I don't like complex clauses but if this bothers mypy, it's probably fine to do what you did for now. Wonder if explicit annotations on the vars would help...

webknjaz commented:
@yuvipanda looks like manual cherry-picking is needed
